builtin/logical/database/dbs/connectionproducer.go

```diff
@@ -0,0 +1,254 @@
+
+	db *sql.DB
+	sync.Mutex
+}
+
+func (cp *sqlConnectionProducer) Connection() (interface{}, error) {
+	// Grab the write lock
+	cp.Lock()
+	defer cp.Unlock()
+
+	// If we already have a DB, we got it!
+	if cp.db != nil {
+		if err := cp.db.Ping(); err == nil {
+			return cp.db, nil
+		}
+		// If the ping was unsuccessful, close it and ignore errors as we'll be
+		// reestablishing anyways
+		cp.db.Close()
+	}
+
+	// Otherwise, attempt to make connection
+	conn := cp.connDetails.ConnectionURL
+
+	// Ensure timezone is set to UTC for all the conenctions
+	if strings.HasPrefix(conn, "postgres://") || strings.HasPrefix(conn, "postgresql://") {
+		if strings.Contains(conn, "?") {
+			conn += "&timezone=utc"
+		} else {
+			conn += "?timezone=utc"
+		}
+	} else {
+		conn += " timezone=utc"
+	}
+
+	var err error
+	cp.db, err = sql.Open(cp.config.DatabaseType, conn)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set some connection pool settings. We don't need much of this,
+	// since the request rate shouldn't be high.
+	cp.db.SetMaxOpenConns(cp.config.MaxOpenConnections)
+	cp.db.SetMaxIdleConns(cp.config.MaxIdleConnections)
+
+	return cp.db, nil
+}
+
+func (cp *sqlConnectionProducer) Close() {
+	// Grab the write lock
+	cp.Lock()
+	defer cp.Unlock()
+
+	if cp.db != nil {
+		cp.db.Close()
+	}
+
+	cp.db = nil
+}
+
+func (cp *sqlConnectionProducer) Reset(config *DatabaseConfig) error {
+	// Grab the write lock
+	cp.Lock()
+
+	var details *sqlConnectionDetails
+	err := mapstructure.Decode(config.ConnectionDetails, &details)
+	if err != nil {
+		return err
+	}
+
+	cp.connDetails = details
+	cp.config = config
+
+	cp.Unlock()
+
+	cp.Close()
+	_, err = cp.Connection()
+	return err
+}
+
+// cassandraConnectionProducer impliments ConnectionProducer and provides connections for cassandra
+type cassandraConnectionDetails struct {
+	Hosts           string `json:"hosts" structs:"hosts" mapstructure:"hosts"`
+	Username        string `json:"username" structs:"username" mapstructure:"username"`
+	Password        string `json:"password" structs:"password" mapstructure:"password"`
+	TLS             bool   `json:"tls" structs:"tls" mapstructure:"tls"`
+	InsecureTLS     bool   `json:"insecure_tls" structs:"insecure_tls" mapstructure:"insecure_tls"`
+	Certificate     string `json:"certificate" structs:"certificate" mapstructure:"certificate"`
+	PrivateKey      string `json:"private_key" structs:"private_key" mapstructure:"private_key"`
+	IssuingCA       string `json:"issuing_ca" structs:"issuing_ca" mapstructure:"issuing_ca"`
+	ProtocolVersion int    `json:"protocol_version" structs:"protocol_version" mapstructure:"protocol_version"`
+	ConnectTimeout  int    `json:"connect_timeout" structs:"connect_timeout" mapstructure:"connect_timeout"`
+	TLSMinVersion   string `json:"tls_min_version" structs:"tls_min_version" mapstructure:"tls_min_version"`
+	Consistancy     string `json:"tls_min_version" structs:"tls_min_version" mapstructure:"tls_min_version"`
```

Comment:
In the old code Consistency was part of the role. This part is still a WIP but i am planning on moving it over to the ConnectionDetails instead.